### DISCUSSION TITLE
✅  Mosh Support

### DISCUSSION THREAD

**Discussion by acbee** [isontheline/pro.webssh.net#380] [Ideas]
https://mosh.org/

**Comment by isontheline** [comment]
Yes but Mosh is released under a GPLv3+ (GNU GPL version 3 or later) license :/
I need a MIT implementation.

**Comment by rnkn** [comment]
Could Apache 2.0 work? https://github.com/MisterTea/EternalTerminal

**Reply by isontheline** [reply]
It could 😉

Doesn't SSH+tmux could be used instead of a completely different protocol?

**Reply by rnkn** [reply]
ssh+tmux solves part of the problem, i.e. persistent session, but not persistent connection. The advantage of mosh is the combination of both.

**Reply by rnkn** [reply]
According to [this page](https://support.termius.com/hc/en-us/articles/4402049882137--Supported-SSH-Capabilities):
“Termius uses its own library for Mosh compatibility.” But it is closed source.

**Reply by isontheline** [reply]
Yes a custom implementation of the mosh client could be a solution but I can't give any ETA as WebSSH have already a lot of unclosed issues 😢

**Reply by rnkn** [reply]
Sounds great! Thanks

**Reply by enveezee** [reply]
Yes it's no small request but many other ssh tools support mosh and mosh is more ideal for the platform because it's udp rather than tcp and survives network hops and such you have on mobile platforms gracefully and it would also address iOS issues with inability to keep alive in the background because mosh doesn't work like ssh does on connect it uses ssh to start a mosh server instance which sends back a key to the client then the client can use this key to communicate with the server as long as the server process lives and it can tolerate infinite lag and loss of direct connectivity and automatically resume when contact is restored. It's certainly worth exploring. I am only here and thus far other than lack of mosh support fairly satisfied though were it not for the fact apples ecosystem is so exclusive and expensive and thus there are few good options for things and that Termius did unthinkable damage to my server resulting in significant data loss I wouldn't even be here or be using this app cause ever since I found mosh it's far preferable to ssh especially when outside the lan and more so on mobile and even more so on a platform like iOS that is hostile to background tasks

**Comment by isontheline** [comment]
@enveezee Which damagees does termius have done?

I will check again if any new Mosh implementation is available (outside GPL)

**Reply by enveezee** [reply]
Well I didn't fully map what precisely happened but I did an SFTP upload to my computer and it somehow after a recent update after they had completely overhauled the UI, had not only deleted everything in the target dir, but also everything in all the sibling dirs and even removed some of the sibling dirs. It didn't go beyond that one level up and thankfully the target dir was two levels deep or it'd have likely deleted everything in my homedir. I immediately removed the app and gave it a bad review and then did what I could as far as data recovery and I been using your app ever since. 

I wasn't aware of the problems with GPL and Apple development. Its yet another reason now why I don't even bother trying to support Apple platforms in my own code, though it makes me more conscious of using a more permissible MIT license for things I write. As far as I know there aren't any more permissible MOSH implementations out there and termius wrote their own to work around this. I had wondered why its so hard to find a MOSH implementation on iOS and now I know. However, while MOSH is GPLv3 the developers have expressly stated exceptions for where iOS is concerned https://github.com/mobile-shell/mosh/blob/master/COPYING.iOS  https://github.com/mobile-shell/mosh/blob/master/ocb-license.html  in that as long as you are compliant with the rest of the provisions provided for in the license, they are waiving this unintended conflict of interest between the GPL terms and those of Apple's terms so I think you're in the clear as they seem to have made it very clear they did not intend for this to happen and have made efforts to make it known that this problem is one Apple created and they didn't intend.

**Reply by isontheline** [reply]
Hey @enveezee 👋

Thanks for sharing all those details about what happened with Termius - that sounds like a really scary data loss situation!  I'm glad you were able to recover at least some of your data and that you've been satisfied with WebSSH since then.

I really appreciate you taking the time to explain the Mosh licensing situation and pointing me to those iOS exception files. You're right that the Mosh developers have made efforts to address the GPL/iOS conflict with specific exceptions.

However, I remain reluctant to integrate GPL code into WebSSH, even with those exceptions. As an indie developer maintaining this app solo, I need to be extremely careful about licensing to protect both myself and the app's future. The GPL license - even with iOS-specific exceptions - creates complexities that could potentially affect how I can license and distribute WebSSH going forward. I prefer to stick with MIT and similarly permissive licenses to keep things clean and straightforward.

I totally understand that Mosh would be a game-changer for mobile SSH usage with its UDP-based approach and resilience to network changes. It's definitely on my radar, and I keep checking periodically if any MIT-licensed implementations emerge. If Termius managed to write their own implementation, it shows it's technically possible - but it's also a massive undertaking for a solo developer with an already long list of issues to tackle.

In the meantime, have you tried using tmux within WebSSH? You can easily start tmux sessions and reattach to them when reconnecting. This solves a lot of the issues that Mosh addresses - your sessions remain persistent on the server side, and you can pick up exactly where you left off even after network interruptions or iOS backgrounding. It won't give you the seamless UDP reconnection that Mosh offers, but it handles the session persistence part really well with no pain!

I'll continue monitoring the situation for Mosh, but I can't promise any timeline for support given the licensing constraints and development priorities.

Have a great day ☀️

Arnaud

**Comment by isontheline** [comment]
I'm thinking about it #1562

**Reply by isontheline** [reply]
mosh protocol is now supported 🎉

No mosh original source code was used, private implementation

Have a great day ☀️
